docs/docs-beta/docs/guides/build/project-structure/creating-a-component.md

```diff
@@ -0,0 +1,149 @@
+
+- `get_schema`: This method should return a Pydantic model that defines the schema for the component. This is the schema for the data that goes into `component.yaml`.
+- `build_defs`: This method should return a `Definitions` object for this component.
+
+
+## Defining a schema
+
+The first step is to define a schema for the component. This means determining what aspects of the component should be customizable.
+
+In this case, we'll want to define a few things:
+
+- The path to the shell script that we'll want to run.
+- The attributes of the asset that we expect this script to produce.
+- Any tags or configuration related to the underlying compute.
+
+To simplify common use cases, `dagster-components` provides schemas for common bits of configuration:
+
+- `AssetAttributesModel`: This contains attributes that are common to all assets, such as the key, description, tags, and dependencies.
+- `OpSpecModel`: This contains attributes specific to an underlying operation, such as the name and tags.
+
+We can the schema for our component and add it to our class as follows:
+
+<CodeExample filePath="guides/components/shell-command/with-config-schema.py" language="python" />
+
+
+## Building definitions
+
+Now that we've defined how the component is parametrized, we need to define how to turn those parameters into a `Definitions` object.
+
+To do so, there are two methods that need to be overridden:
+
+- `load`: This method is responsible for loading the configuration from the `component.yaml` file and translating it into the corresponding Python class.
+- `build_defs`: This method is responsible for returning a `Definitions` object.
+
+In our case, our `load` method will just instantiate our class with a copy of the loaded parameters without modification.
+
+Our `build_defs` method will create a single `@asset` that executes the provided shell script. By convention, we'll put the code to actually execute this asset inside of a function called `evaluate`. This makes it easier for future developers to create subclasses of this component.
+
+<CodeExample filePath="guides/components/shell-command/with-build-defs.py" language="python" />
+
+## Component registration
+
+Following the steps above will automatically register your component type in your environment. You can now run:
+
+```bash
+dg component-type list
+```
+
+and see your new component type in the list of available component types.
+
+## [Advanced] Custom templating
+
+The components system supports a rich templating syntax that allows you to load arbitrary python values based off of your templated `component.yaml` file.
+
+When creating the schema for your component, you can specify custom output types that should be rendered from
+
+### Defining a resolvable field
+
+When creating a schema for your component, if you have a field that should have some custom resolution logic, you can annotate that field with the `ResolvableFieldInfo` class. This allows you to specify:
```

Comment:
Motivating this section with a concrete example might be nice, rather than `complex_obj1`. I wonder if you could continue building on the shell script example, e.g provide a scope for the `code_location_dir` or something?

Reply:
yeah motiviating example is a good idea -- the other part I wanted to capture was the fact that this handles arbitary python object, lmk if this feels better to you